hydradx/model/amm/centralized_market.py

```diff
@@ -0,0 +1,174 @@
+import copy
+from .agents import Agent
+from .amm import AMM
+
+
+class OrderBook:
+    def __init__(self, bids: list[list], asks: list[list[float]]):
+        """
+        bids and asks are in the form of (price: float, quantity: float) tuples
+        could add an ID in there later
+        """
+        self.bids = bids
+        self.asks = asks
+
+    def copy(self):
+        return copy.deepcopy(self)
+
+
+# faster
+OrderBook.copy = lambda self: OrderBook(
+    bids=[list(i for i in bid) for bid in self.bids],
+    asks=[list(i for i in ask) for ask in self.asks],
+)
+
+
+class CentralizedMarket(AMM):
+    def __init__(
+            self,
+            order_book: dict[tuple[str, str], OrderBook],
+            asset_list: list[str] = None,
+            trade_fee: float = 0
+    ):
+        super().__init__()
+        self.order_book = order_book
+        if asset_list:
+            self.asset_list = asset_list
+        else:
+            # using a dict instead of a set here to preserve order (python 3.7+)
+            self.asset_list = list({tkn: 0 for pair in self.order_book for tkn in pair}.keys())
+        if 'USD' not in self.asset_list:
+            self.asset_list = ['USD', *self.asset_list]
+        self.trade_fee = trade_fee
+
+    def swap(
+        self,
+        agent: Agent,
+        tkn_sell: str,
+        tkn_buy: str,
+        buy_quantity: float = 0,
+        sell_quantity: float = 0
+    ):
+        if sell_quantity > agent.holdings[tkn_sell]:
+            raise AssertionError('Agent does not have enough holdings to execute trade.')
+        if tkn_sell not in self.asset_list or tkn_buy not in self.asset_list:
+            raise AssertionError('Asset not found in centralized market.')
+
+        if self.asset_list.index(tkn_sell) < self.asset_list.index(tkn_buy):
```

Comment:
This isn't going to support all asset pairs, we're going to need to be able to support arbitrary asset pairs, i.e. whatever choices any order books happen to make.